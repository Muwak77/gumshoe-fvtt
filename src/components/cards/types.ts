export type CardsSortOrder = "atoz" | "ztoa" | "newest" | "oldest";
export type CardsViewMode = "compact" | "expanded";

export type CardsAreaSettings = {
  category: string;
  sortOrder: CardsSortOrder;
  viewMode: CardsViewMode;
};
