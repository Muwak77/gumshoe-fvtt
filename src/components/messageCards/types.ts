import * as constants from "../../constants";

type AbilityCardModeTestSpend =
  | typeof constants.htmlDataModeTest
  | typeof constants.htmlDataModeSpend
  | typeof constants.htmlDataModeAttack;

export type AbilityCardMode =
  | AbilityCardModeTestSpend
  | typeof constants.htmlDataModeAttack
  | typeof constants.htmlDataModeMwTest
  | typeof constants.htmlDataModeMwNegate
  | typeof constants.htmlDataModeMwWallop;

export const isAbilityCardMode = (
  candidate: string,
): candidate is AbilityCardMode => {
  return (
    candidate === constants.htmlDataModeTest ||
    candidate === constants.htmlDataModeSpend ||
    candidate === constants.htmlDataModeAttack ||
    candidate === constants.htmlDataModeMwTest ||
    candidate === constants.htmlDataModeMwNegate ||
    candidate === constants.htmlDataModeMwWallop
  );
};

export interface MWResult {
  text: string;
  color: string;
}
