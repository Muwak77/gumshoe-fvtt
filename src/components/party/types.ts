import { AbilityType } from "../../types";
import { AbilityItem } from "../../v10Types";

export const typeHeaderKey = "typeHeader" as const;
export const categoryHeaderKey = "categoryHeader" as const;
export const abilityRowKey = "abilityRowString" as const;
export type TypeHeaderData = {
  rowType: typeof typeHeaderKey;
  abilityType: AbilityType;
};
export type CategoryHeaderData = {
  rowType: typeof categoryHeaderKey;
  category: string;
};
export type ActorAbilityInfo = {
  actorId: string;
  abilityId: string | undefined;
  rating: number | undefined;
};
export type AbilityRowData = {
  rowType: typeof abilityRowKey;
  abilityItem: AbilityItem;
  actorInfo: {
    [actorId: string]: ActorAbilityInfo;
  };
  total: number;
};
export type RowData = TypeHeaderData | CategoryHeaderData | AbilityRowData;
export const isTypeHeader = (data: RowData): data is TypeHeaderData =>
  data.rowType === typeHeaderKey;
export const isCategoryHeader = (data: RowData): data is CategoryHeaderData =>
  data.rowType === categoryHeaderKey;
