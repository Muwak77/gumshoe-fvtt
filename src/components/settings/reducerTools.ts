import { Draft, produce } from "immer";
import React, {
  Context,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";

/**
 * A minimal case for all actions - there will always be a `type` and a
 * `payload` property, but the payload may be `undefined`.
 */
export type AnyAction = {
  type: string;
  payload?: unknown;
};

/**
 * A minimal reimplementation of the `createAction` function from
 * `@reduxjs/toolkit`.
 *
 * Given a unique string (danger will robinson!) and a reducer function, returns
 * an object with
 *
 *  * a `create` method that returns an action object
 *  * an `apply` function, which take a state and an action and applies the
 *    relevant reducer if the action matches the case
 */
function createCase<S, P = void>(
  type: string,
  reducer: (draft: S, payload: P) => void,
) {
  const create = (payload: P) => ({ type, payload });
  const match = (action: AnyAction): action is { type: string; payload: P } =>
    action.type === type;
  const apply = (state: S, action: AnyAction) => {
    if (match(action)) {
      return reducer ? reducer(state, action.payload) : state;
    }
    return state;
  };
  return {
    /** Create an action (to be dispatched with `dispatch` from `useReducer` */
    create,
    /**
     * given a mutable draft object of S, and action, mutate the draft if the
     * action is from this slice.
     */
    apply,
  };
}

type Reducers<S> = { [key: string]: (state: Draft<S>, payload?: any) => void };

export type CreateSliceArgs = {
  onError?: (error: unknown) => void;
  log?: (...args: unknown[]) => void;
};

/**
 * A minimal reimagination of the `createSlice` function from
 * `@reduxjs/toolkit`.
 * This function is curried so that we can specify S (state) but allow R
 * (reducers) to be inferred.
 */
export const createSlice =
  <S extends object>(args: CreateSliceArgs = {}) =>
  <R extends Reducers<S>>(reducers: R) => {
    // turn all the reducers into cases (so they can be created and applied)
    const sliceCases = Object.entries(reducers).map(([key, reducer]) => {
      const action = createCase(key, reducer);
      return [key, action] as const;
    });
    // turn those cases back into an object of creators.
    const creators = Object.fromEntries(
      sliceCases.map(([key, action]) => [key, action.create]),
    ) as {
      // we need this mad cast because `map` can't give us accurate types.
      // we're just knitting together the `typeof createCase<...>["create"]`
      // and working out the payload type using inference.
      [key in keyof R]: ReturnType<
        typeof createCase<
          S,
          R[key] extends (state: infer S1) => void
            ? void
            : R[key] extends (state: infer S1, payload: infer P1) => void
              ? P1
              : never
        >
      >["create"];
    };
    // `useDispatch` takes the initial state as an argument, so we don't need to
    // provide a default argument for `state` here (unlike e.g. a Redux
    // reducer).
    const reducer = (state: S, action: AnyAction) => {
      try {
        const newState = produce(state, (draft) => {
          for (const [, sliceCase] of sliceCases) {
            sliceCase.apply(draft, action);
          }
        });
        if (args.log) {
          args.log(action, state, newState);
        }
        return newState;
      } catch (e) {
        // in the event of an error, we will try to keep going rather than just
        // exploding
        if (args.onError) {
          args.onError(e);
        }
        return state;
      }
    };

    return {
      /**
       * A dictionary of action creators, keyed by the name of the case.
       */
      creators,
      /**
       * The reducer function for this slice.
       */
      reducer,
    };
  };

/**
 * Create a `useSelector` hook that uses the given `context`.
 */
export const createUseSelectorHook = <TState>(context: Context<TState>) => {
  return <TValue>(selector: (state: TState) => TValue) => {
    const state = useContext(context);
    const [value, setValue] = useState<TValue>(selector(state));
    const frozenRef = useRef(false);
    useEffect(() => {
      if (!frozenRef.current) {
        setValue(selector(state));
      }
    }, [selector, state]);
    const freeze = useCallback(() => {
      frozenRef.current = true;
    }, []);
    return { value, freeze };
  };
};
