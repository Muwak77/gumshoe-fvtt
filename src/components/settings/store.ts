import {
  CardCategory,
  EquipmentFieldMetadata,
  PresetV1,
} from "@lumphammer/investigator-fvtt-types";
import { nanoid } from "nanoid";

import {
  getDevMode,
  moveKeyDown,
  moveKeyUp,
  renameProperty,
  systemLogger,
} from "../../functions/utilities";
import { pathOfCthulhuPreset } from "../../presets";
import { SettingsDict } from "../../settings/settings";
import { EquipmentFieldType } from "../../types";
import { assertNumericFieldOkayness } from "./functions";
import { createSlice, CreateSliceArgs } from "./reducerTools";
import { PcOrNpc, State } from "./types";

const defaultStoreArgs: CreateSliceArgs = {
  log(...args) {
    if (getDevMode()) {
      systemLogger.log(...args);
    }
  },
  onError(e) {
    systemLogger.error("Reducer error", e);
    if (ui && e instanceof Error) {
      ui.notifications?.error(`Settings error: ${e.toString()}`, {
        permanent: true,
      });
    }
  },
};

export const createSystemSlice = (args: CreateSliceArgs) =>
  createSlice<State>(args)({
    setSome: (draft, payload: { newSettings: Partial<SettingsDict> }) => {
      Object.assign(draft.settings, payload.newSettings);
    },
    addCategory: (draft: State) => {
      draft.settings.equipmentCategories[nanoid()] = {
        name: "New category",
        fields: {},
      };
    },
    deleteCategory: (
      { settings: { equipmentCategories } }: State,
      { id }: { id: string },
    ) => {
      delete equipmentCategories[id];
    },
    renameCategory: (
      { settings: { equipmentCategories: cats } }: State,
      payload: { id: string; newName: string },
    ) => {
      cats[payload.id].name = payload.newName;
    },
    changeCategoryId: (
      { settings }: State,
      payload: { oldCategoryId: string; newCategoryId: string },
    ) => {
      const newCats: typeof settings.equipmentCategories = {};
      if (settings.equipmentCategories[payload.newCategoryId]) {
        throw new Error(
          `Cannot change category id to "${payload.newCategoryId}" - already exists`,
        );
      }
      for (const [id, category] of Object.entries(
        settings.equipmentCategories,
      )) {
        if (id === payload.oldCategoryId) {
          newCats[payload.newCategoryId] = category;
        } else {
          newCats[id] = category;
        }
      }
      settings.equipmentCategories = newCats;
    },
    moveCategoryUp: (
      { settings }: State,
      { categoryId }: { categoryId: string },
    ) => {
      settings.equipmentCategories = moveKeyUp(
        settings.equipmentCategories,
        categoryId,
      );
    },
    moveCategoryDown: (
      { settings }: State,
      { categoryId }: { categoryId: string },
    ) => {
      settings.equipmentCategories = moveKeyDown(
        settings.equipmentCategories,
        categoryId,
      );
    },
    addField: (
      { settings: { equipmentCategories: cats } }: State,
      payload: { categoryId: string },
    ) => {
      cats[payload.categoryId].fields[nanoid()] = {
        name: "New Field",
        type: "string",
        default: "",
      };
    },
    deleteField: (
      { settings: { equipmentCategories: cats } }: State,
      payload: { categoryId: string; fieldId: string },
    ) => {
      delete cats[payload.categoryId].fields[payload.fieldId];
    },
    renameField: (
      { settings }: State,
      payload: { categoryId: string; fieldId: string; newName: string },
    ) => {
      settings.equipmentCategories[payload.categoryId].fields[
        payload.fieldId
      ].name = payload.newName;
    },
    changeFieldId: (
      { settings: { equipmentCategories: cats } }: State,
      payload: { categoryId: string; fieldId: string; newFieldId: string },
    ) => {
      const newFields: Record<string, EquipmentFieldMetadata> = {};
      if (cats[payload.categoryId].fields[payload.newFieldId]) {
        throw new Error(
          `Cannot change field id to "${payload.newFieldId}" - already exists`,
        );
      }
      for (const [id, field] of Object.entries(
        cats[payload.categoryId].fields,
      )) {
        if (id === payload.fieldId) {
          newFields[payload.newFieldId] = field;
        } else {
          newFields[payload.fieldId] = field;
        }
      }
      cats[payload.categoryId].fields = newFields;
    },

    setFieldType: (
      { settings: { equipmentCategories: cats } }: State,
      payload: {
        categoryId: string;
        fieldId: string;
        newType: EquipmentFieldType;
      },
    ) => {
      const field = cats[payload.categoryId].fields[payload.fieldId];
      field.type = payload.newType;
      if (payload.newType === "number") {
        field.default = 0;
      } else if (payload.newType === "checkbox") {
        field.default = false;
      } else if (payload.newType === "string") {
        field.default = "";
      }
    },
    setFieldDefault: (
      { settings: { equipmentCategories: cats } }: State,
      payload: {
        categoryId: string;
        fieldId: string;
        newDefault: string | number | boolean;
      },
    ) => {
      const field = cats[payload.categoryId].fields[payload.fieldId];
      if (
        (field.type === "string" && typeof payload.newDefault !== "string") ||
        (field.type === "number" && typeof payload.newDefault !== "number") ||
        (field.type === "checkbox" && typeof payload.newDefault !== "boolean")
      ) {
        throw new Error(
          `Invalid default value ${payload.newDefault} for field type ${field.type}`,
        );
      }
      field.default = payload.newDefault;
    },
    setFieldMin: (
      { settings: { equipmentCategories: cats } }: State,
      payload: {
        categoryId: string;
        fieldId: string;
        newMin: number | undefined;
      },
    ) => {
      const field = cats[payload.categoryId].fields[payload.fieldId];
      assertNumericFieldOkayness(field, payload.fieldId, payload.newMin);
      if (
        field.max !== undefined &&
        payload.newMin !== undefined &&
        payload.newMin > field.max
      ) {
        throw new Error("Min cannot be greater than max");
      }
      field.min = payload.newMin;
    },
    setFieldMax: (
      { settings: { equipmentCategories: cats } }: State,
      payload: {
        categoryId: string;
        fieldId: string;
        newMax: number | undefined;
      },
    ) => {
      const field = cats[payload.categoryId].fields[payload.fieldId];
      assertNumericFieldOkayness(field, payload.fieldId, payload.newMax);
      if (
        field.min !== undefined &&
        payload.newMax !== undefined &&
        payload.newMax < field.min
      ) {
        throw new Error("Max cannot be less than min");
      }
      field.max = payload.newMax;
    },
    moveFieldUp: (
      { settings: { equipmentCategories: cats } }: State,
      { categoryId, fieldId }: { categoryId: string; fieldId: string },
    ) => {
      cats[categoryId].fields = moveKeyUp(cats[categoryId].fields, fieldId);
    },
    moveFieldDown: (
      { settings: { equipmentCategories: cats } }: State,
      { categoryId, fieldId }: { categoryId: string; fieldId: string },
    ) => {
      cats[categoryId].fields = moveKeyDown(cats[categoryId].fields, fieldId);
    },
    applyPreset: (
      draft: State,
      payload: { preset: PresetV1; presetId: string },
    ) => {
      draft.settings = {
        // start with the current temp settings - this way we keep any values
        // not handled by the presets
        ...draft.settings,
        // now we layer in a safe default. This is typed as Required<> so it
        // will always contain one of everything that can be controlled by a
        // preset.
        ...pathOfCthulhuPreset,
        // now layer in the actual preset
        ...payload.preset,
        // and finally, set the actual preset id
        systemPreset: payload.presetId,
      };
    },
    addStat: (draft: State, { which }: { which: PcOrNpc }) => {
      const newId = `stat${Object.keys(draft.settings[which]).length}`;
      draft.settings[which][newId] = { name: "", default: 0 };
    },
    setStatMin: (
      draft: State,
      {
        which,
        statId,
        newMin,
      }: { which: PcOrNpc; statId: string; newMin: number | undefined },
    ) => {
      const max = draft.settings[which][statId].max;
      if (max !== undefined && newMin !== undefined && newMin > max) {
        throw new Error("Min cannot be greater than max");
      }
      draft.settings[which][statId].min = newMin;
    },
    setStatMax: (
      draft: State,
      {
        which,
        statId,
        newMax,
      }: { which: PcOrNpc; statId: string; newMax: number | undefined },
    ) => {
      const min = draft.settings[which][statId].min;
      if (min !== undefined && newMax !== undefined && newMax < min) {
        throw new Error("Max cannot be less than min");
      }
      draft.settings[which][statId].max = newMax;
    },
    setStatDefault: (
      draft: State,
      {
        which,
        statId,
        newDefault,
      }: { which: PcOrNpc; statId: string; newDefault: number },
    ) => {
      draft.settings[which][statId].default = newDefault;
    },
    setStatName: (
      draft: State,
      {
        which,
        statId,
        newName,
      }: { which: PcOrNpc; statId: string; newName: string },
    ) => {
      draft.settings[which][statId].name = newName;
    },
    deleteStat: (
      draft: State,
      { which, statId }: { which: PcOrNpc; statId: string },
    ) => {
      delete draft.settings[which][statId];
    },
    setStatId: (
      draft: State,
      {
        which,
        oldStatId,
        newStatId,
      }: { which: PcOrNpc; oldStatId: string; newStatId: string },
    ) => {
      draft.settings[which] = renameProperty(
        oldStatId,
        newStatId,
        draft.settings[which],
      );
    },
    throwError: (draft: State, { message }: { message: string }) => {
      throw new Error(message);
    },
    addCardCategory: (draft: State, payload: { id: string }) => {
      draft.settings.cardCategories.push({
        id: payload.id,
        singleName: "New category",
        pluralName: "New category",
      });
    },
    renameCardCategory: (
      draft: State,
      { id, newName }: { id: string; newName: string },
    ) => {
      draft.settings.cardCategories.find((c) => c.id === id)!.singleName =
        newName;
    },
    deleteCardCategory: (draft: State, { id }: { id: string }) => {
      draft.settings.cardCategories = draft.settings.cardCategories.filter(
        (c) => c.id !== id,
      );
    },
    setCardCategoryCssClass: (
      draft: State,
      { id, newCssClass }: { id: string; newCssClass: string },
    ) => {
      draft.settings.cardCategories.find((c) => c.id === id)!.cssClass =
        newCssClass;
    },
    setCardCategoryId: (
      draft: State,
      { id, newId }: { id: string; newId: string },
    ) => {
      draft.settings.cardCategories.find((c) => c.id === id)!.id = newId;
    },
    setCardCategories: (
      draft: State,
      { newCardCategories }: { newCardCategories: CardCategory[] },
    ) => {
      draft.settings.cardCategories = newCardCategories;
    },
  });

export const store = createSystemSlice(defaultStoreArgs);
