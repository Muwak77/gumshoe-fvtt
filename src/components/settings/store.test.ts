// import { nanoid } from "nanoid";
import { createSystemSlice } from "./store";
import { State } from "./types";
import { expect, jest, it, describe } from "@jest/globals";
// import * as functions from "../../functions";
import { AnyAction } from "./reducerTools";
import { pathOfCthulhuPreset } from "../../presets";
// import { createPatch } from "diff";
import { diff } from "jest-diff";

type TestTuple = [string, AnyAction];

const initialState: State = {
  settings: {
    pcStats: {
      pcStat0: {
        min: 0,
        max: 10,
        default: 3,
        name: "PC Statisitic Zero",
      },
    },
    npcStats: {
      npcStat0: {
        min: 0,
        max: 10,
        default: 3,
        name: "NPC Statisitic Zero",
      },
    },
    systemPreset: "dnd5e",
    combatAbilities: [],
    customThemePath: "",
    equipmentCategories: {
      category0: {
        name: "initial category",
        fields: {
          field0: {
            name: "Initial field",
            type: "string",
            default: "default value",
          },
          field1: {
            name: "Initial field",
            type: "string",
            default: "default value",
          },
        },
      },
      category1: {
        name: "initial category",
        fields: {},
      },
    },
    debugTranslations: false,
    useBoost: false,
    defaultThemeName: "",
    useMwInjuryStatus: false,
    useMwStyleAbilities: false,
    useNpcCombatBonuses: false,
    generalAbilityCategories: [],
    migrationFlags: {
      actor: {},
      item: {},
      compendium: {},
      journal: {},
      macro: {},
      scene: {},
      rollTable: {},
      playlist: {},
      world: {},
    },
    useTurnPassingInitiative: false,
    genericOccupation: "",
    investigativeAbilityCategories: [],
    longNotes: [],
    mwHiddenShortNotes: [],
    mwUseAlternativeItemTypes: false,
    newNPCPacks: [],
    newPCPacks: [],
    occupationLabel: "",
    personalDetails: [],
    showEmptyInvestigativeCategories: false,
    systemMigrationVersion: "",
    abilityCategories: "",
    shortNotes: [],
  },
};

// const initialStateJSON = JSON.stringify(initialState, null, 2);

const noColor = (x: string) => x;

const slice = createSystemSlice({
  log: jest.fn(),
  onError: jest.fn(),
});

// we need to mock nanoid so that we can predictably generate ids
// I tried putting this in beforeAll but for some reason we weren't seeing the
// mock in the test
jest.mock("nanoid", () => {
  let count = 0;
  return {
    __esModule: true,
    nanoid: () => "id" + count++,
  };
});

afterAll(() => {
  jest.resetModules();
});

/*
 * DIFF SNAPSHOT TESTS!
 *
 * Rather than maintain a manual list of expected values,
 * we can just snapshot the diff between the initial state and the state after
 * the action is applied. This means the tests are not super brittle in the
 * event of the structure changing, but it does require you to manually check
 * the diff to make sure it's what you expect.
 *
 * The snapshots are stored in the __snapshots__ folder.
 *
 * To update the snapshots, run `npx jest-u`
 *
 * AND PLEASE CHECK THE DIFFS AFTER UPDATING! Your git tool of choice will help
 * here, so you can just compare the diffs that have changed. Yes, it might feel
 * a bit weird looking at "diffs of diffs", but you'll get the hang of it.
 */
describe("reducer", () => {
  it("should ignore unknown actions", () => {
    const result = slice.reducer(initialState, { type: "FOOEY" });
    // no point snapshotting this
    expect(result).toEqual(initialState);
  });

  it.each<TestTuple>([
    [
      "set a value using setSome",
      slice.creators.setSome({ newSettings: { genericOccupation: "foo" } }),
    ],
    ["add a category", slice.creators.addCategory()],
    ["delete a category", slice.creators.deleteCategory({ id: "category0" })],
    [
      "rename a category",
      slice.creators.renameCategory({ id: "category0", newName: "foo" }),
    ],
    [
      "change a category id",
      slice.creators.changeCategoryId({ oldId: "category0", newId: "foo" }),
    ],
    ["move a category up", slice.creators.moveCategoryUp({ id: "category1" })],
    [
      "move a category down",
      slice.creators.moveCategoryDown({ id: "category0" }),
    ],
    ["add a field", slice.creators.addField({ categoryId: "category0" })],
    [
      "delete a field",
      slice.creators.deleteField({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "rename a field",
      slice.creators.renameField({
        categoryId: "category0",
        fieldId: "field0",
        newName: "foo",
      }),
    ],
    [
      "change a field id",
      slice.creators.changeFieldId({
        categoryId: "category0",
        fieldId: "field0",
        newFieldId: "foo",
      }),
    ],
    [
      "set a field type",
      slice.creators.setFieldType({
        categoryId: "category0",
        fieldId: "field0",
        newType: "number",
      }),
    ],
    [
      "set a field default",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field0",
        newDefault: "foo",
      }),
    ],
    [
      "set a field minimum",
      slice.creators.setFieldMin({
        categoryId: "category0",
        fieldId: "field0",
        newMin: 1,
      }),
    ],
    [
      "set a field maximum",
      slice.creators.setFieldMax({
        categoryId: "category0",
        fieldId: "field0",
        newMax: 1,
      }),
    ],
    [
      "apply a preset",
      slice.creators.applyPreset({
        preset: pathOfCthulhuPreset,
        presetId: "pathOfCthulhu",
      }),
    ],
    [
      "add a stat",
      slice.creators.addStat({
        which: "pcStats",
      }),
    ],
    [
      "set a stat minimum",
      slice.creators.setStatMin({
        which: "pcStats",
        id: "pcStat0",
        value: 1,
      }),
    ],
    [
      "set a stat maximum",
      slice.creators.setStatMax({
        which: "pcStats",
        id: "pcStat0",
        value: 1,
      }),
    ],
    [
      "move a field up",
      slice.creators.moveFieldUp({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "move a field down",
      slice.creators.moveFieldDown({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "set a stat default",
      slice.creators.setStatDefault({
        which: "pcStats",
        id: "pcStat0",
        value: 1,
      }),
    ],
    [
      "set a stat name",
      slice.creators.setStatName({
        which: "pcStats",
        id: "pcStat0",
        name: "foo",
      }),
    ],
    [
      "delete a stat",
      slice.creators.deleteStat({
        which: "pcStats",
        id: "pcStat0",
      }),
    ],
    [
      "set a stat id",
      slice.creators.setStatId({
        which: "pcStats",
        oldId: "pcStat0",
        newId: "foo",
      }),
    ],
  ])("should %s", (name, action) => {
    // * why snapshots of diffs?
    //   * they're smaller and easier to read than snapshotting the entire
    //     output object. You can peruse the snapshots file and see the
    //     *changes* caused by the action, rather than the entire state.
    // * why a patch formatted diff?
    //   * first attempt used an object diff (from `just-diff`) but that does
    //     not detect changes in key order.
    const result = slice.reducer(initialState, action);
    // const resultJSON = JSON.stringify(result, null, 2);
    const diffs = diff(initialState, result, {
      aColor: noColor,
      bColor: noColor,
      changeColor: noColor,
      commonColor: noColor,
      patchColor: noColor,
      contextLines: 3,
      expand: false,
      omitAnnotationLines: true,
    });
    expect(diffs).toMatchSnapshot();
  });
});
