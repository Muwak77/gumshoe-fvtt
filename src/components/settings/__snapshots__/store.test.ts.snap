// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should add a category 1`] = `
"@@ -25,6 +25,10 @@
          "fields": Object {},
          "name": "initial category",
        },
+       "id0": Object {
+         "fields": Object {},
+         "name": "New category",
+       },
      },
      "generalAbilityCategories": Array [],
      "genericOccupation": "","
`;

exports[`reducer should add a field 1`] = `
"@@ -18,6 +18,11 @@
              "name": "Initial field",
              "type": "string",
            },
+           "id1": Object {
+             "default": "",
+             "name": "New Field",
+             "type": "string",
+           },
          },
          "name": "initial category",
        },"
`;

exports[`reducer should add a stat 1`] = `
"@@ -61,6 +61,10 @@
          "min": 0,
          "name": "PC Statisitic Zero",
        },
+       "stat1": Object {
+         "default": 0,
+         "name": "",
+       },
      },
      "personalDetails": Array [],
      "shortNotes": Array [],"
`;

exports[`reducer should apply a preset 1`] = `
"@@ -1,35 +1,37 @@
  Object {
    "settings": Object {
      "abilityCategories": "",
-     "combatAbilities": Array [],
+     "combatAbilities": Array [
+       "Scuffling",
+       "Weapons",
+       "Firearms",
+       "Athletics",
+     ],
      "customThemePath": "",
      "debugTranslations": false,
-     "defaultThemeName": "",
+     "defaultThemeName": "tealTheme",
+     "displayName": "Path of Cthulhu",
      "equipmentCategories": Object {
-       "category0": Object {
-         "fields": Object {
-           "field0": Object {
-             "default": "default value",
-             "name": "Initial field",
-             "type": "string",
-           },
-           "field1": Object {
-             "default": "default value",
-             "name": "Initial field",
-             "type": "string",
-           },
-         },
-         "name": "initial category",
-       },
-       "category1": Object {
+       "general": Object {
          "fields": Object {},
-         "name": "initial category",
+         "name": "General",
        },
      },
-     "generalAbilityCategories": Array [],
-     "genericOccupation": "",
-     "investigativeAbilityCategories": Array [],
-     "longNotes": Array [],
+     "generalAbilityCategories": Array [
+       "General",
+     ],
+     "genericOccupation": "Investigator",
+     "investigativeAbilityCategories": Array [
+       "Academic",
+       "Interpersonal",
+       "Technical",
+     ],
+     "longNotes": Array [
+       "Notes, Contacts etc.",
+       "Occupational Benefits",
+       "Pillars of Sanity",
+       "Sources of Stability",
+     ],
      "migrationFlags": Object {
        "actor": Object {},
        "compendium": Object {},
@@ -43,30 +45,52 @@
      },
      "mwHiddenShortNotes": Array [],
      "mwUseAlternativeItemTypes": false,
-     "newNPCPacks": Array [],
-     "newPCPacks": Array [],
+     "newNPCPacks": Array [
+       "investigator.opponentAbilities",
+     ],
+     "newPCPacks": Array [
+       "investigator.pathOfCthulhuAbilities",
+     ],
      "npcStats": Object {
-       "npcStat0": Object {
+       "alertness": Object {
+         "default": 0,
+         "name": "Alertness",
+       },
+       "armor": Object {
+         "default": 0,
+         "name": "Armor",
+       },
+       "hitThreshold": Object {
          "default": 3,
-         "max": 10,
-         "min": 0,
-         "name": "NPC Statisitic Zero",
+         "name": "Hit threshold",
+       },
+       "stabilityLoss": Object {
+         "default": 0,
+         "name": "Stability Loss",
+       },
+       "stealth": Object {
+         "default": 0,
+         "name": "Stealth",
        },
      },
-     "occupationLabel": "",
+     "occupationLabel": "Occupation",
      "pcStats": Object {
-       "pcStat0": Object {
+       "hitThreshold": Object {
          "default": 3,
-         "max": 10,
-         "min": 0,
-         "name": "PC Statisitic Zero",
+         "name": "Hit threshold",
+       },
      },
+     "personalDetails": Array [
+       Object {
+         "name": "Drive",
+         "type": "item",
        },
-     "personalDetails": Array [],
+     ],
+     "schemaVersion": "v1",
      "shortNotes": Array [],
      "showEmptyInvestigativeCategories": false,
      "systemMigrationVersion": "",
-     "systemPreset": "dnd5e",
+     "systemPreset": "pathOfCthulhu",
      "useBoost": false,
      "useMwInjuryStatus": false,
      "useMwStyleAbilities": false,"
`;

exports[`reducer should change a category id 1`] = `
"@@ -6,7 +6,11 @@
      "debugTranslations": false,
      "defaultThemeName": "",
      "equipmentCategories": Object {
-       "category0": Object {
+       "category1": Object {
+         "fields": Object {},
+         "name": "initial category",
+       },
+       "foo": Object {
          "fields": Object {
            "field0": Object {
              "default": "default value",
@@ -19,10 +23,6 @@
              "type": "string",
            },
          },
-         "name": "initial category",
-       },
-       "category1": Object {
-         "fields": Object {},
          "name": "initial category",
        },
      },"
`;

exports[`reducer should change a field id 1`] = `
"@@ -13,7 +13,7 @@
              "name": "Initial field",
              "type": "string",
            },
-           "field1": Object {
+           "foo": Object {
              "default": "default value",
              "name": "Initial field",
              "type": "string","
`;

exports[`reducer should delete a category 1`] = `
"@@ -6,21 +6,6 @@
      "debugTranslations": false,
      "defaultThemeName": "",
      "equipmentCategories": Object {
-       "category0": Object {
-         "fields": Object {
-           "field0": Object {
-             "default": "default value",
-             "name": "Initial field",
-             "type": "string",
-           },
-           "field1": Object {
-             "default": "default value",
-             "name": "Initial field",
-             "type": "string",
-           },
-         },
-         "name": "initial category",
-       },
        "category1": Object {
          "fields": Object {},
          "name": "initial category","
`;

exports[`reducer should delete a field 1`] = `
"@@ -8,11 +8,6 @@
      "equipmentCategories": Object {
        "category0": Object {
          "fields": Object {
-           "field0": Object {
-             "default": "default value",
-             "name": "Initial field",
-             "type": "string",
-           },
            "field1": Object {
              "default": "default value",
              "name": "Initial field","
`;

exports[`reducer should delete a stat 1`] = `
"@@ -54,14 +54,7 @@
        },
      },
      "occupationLabel": "",
-     "pcStats": Object {
-       "pcStat0": Object {
-         "default": 3,
-         "max": 10,
-         "min": 0,
-         "name": "PC Statisitic Zero",
-       },
-     },
+     "pcStats": Object {},
      "personalDetails": Array [],
      "shortNotes": Array [],
      "showEmptyInvestigativeCategories": false,"
`;

exports[`reducer should move a category down 1`] = `"Compared values have no visual difference."`;

exports[`reducer should move a category up 1`] = `"Compared values have no visual difference."`;

exports[`reducer should move a field down 1`] = `"Compared values have no visual difference."`;

exports[`reducer should move a field up 1`] = `"Compared values have no visual difference."`;

exports[`reducer should rename a category 1`] = `
"@@ -19,7 +19,7 @@
              "type": "string",
            },
          },
-         "name": "initial category",
+         "name": "foo",
        },
        "category1": Object {
          "fields": Object {},"
`;

exports[`reducer should rename a field 1`] = `
"@@ -10,7 +10,7 @@
          "fields": Object {
            "field0": Object {
              "default": "default value",
-             "name": "Initial field",
+             "name": "foo",
              "type": "string",
            },
            "field1": Object {"
`;

exports[`reducer should set a field default 1`] = `
"@@ -9,7 +9,7 @@
        "category0": Object {
          "fields": Object {
            "field0": Object {
-             "default": "default value",
+             "default": "foo",
              "name": "Initial field",
              "type": "string",
            },"
`;

exports[`reducer should set a field maximum 1`] = `"Compared values have no visual difference."`;

exports[`reducer should set a field minimum 1`] = `"Compared values have no visual difference."`;

exports[`reducer should set a field type 1`] = `
"@@ -9,9 +9,9 @@
        "category0": Object {
          "fields": Object {
            "field0": Object {
-             "default": "default value",
+             "default": 0,
              "name": "Initial field",
-             "type": "string",
+             "type": "number",
            },
            "field1": Object {
              "default": "default value","
`;

exports[`reducer should set a stat default 1`] = `
"@@ -56,7 +56,7 @@
      "occupationLabel": "",
      "pcStats": Object {
        "pcStat0": Object {
-         "default": 3,
+         "default": 1,
          "max": 10,
          "min": 0,
          "name": "PC Statisitic Zero","
`;

exports[`reducer should set a stat id 1`] = `
"@@ -55,7 +55,7 @@
      },
      "occupationLabel": "",
      "pcStats": Object {
-       "pcStat0": Object {
+       "foo": Object {
          "default": 3,
          "max": 10,
          "min": 0,"
`;

exports[`reducer should set a stat maximum 1`] = `
"@@ -57,7 +57,7 @@
      "pcStats": Object {
        "pcStat0": Object {
          "default": 3,
-         "max": 10,
+         "max": 1,
          "min": 0,
          "name": "PC Statisitic Zero",
        },"
`;

exports[`reducer should set a stat minimum 1`] = `
"@@ -58,7 +58,7 @@
        "pcStat0": Object {
          "default": 3,
          "max": 10,
-         "min": 0,
+         "min": 1,
          "name": "PC Statisitic Zero",
        },
      },"
`;

exports[`reducer should set a stat name 1`] = `
"@@ -59,7 +59,7 @@
          "default": 3,
          "max": 10,
          "min": 0,
-         "name": "PC Statisitic Zero",
+         "name": "foo",
        },
      },
      "personalDetails": Array [],"
`;

exports[`reducer should set a value using setSome 1`] = `
"@@ -27,7 +27,7 @@
        },
      },
      "generalAbilityCategories": Array [],
-     "genericOccupation": "",
+     "genericOccupation": "foo",
      "investigativeAbilityCategories": Array [],
      "longNotes": Array [],
      "migrationFlags": Object {"
`;
